File: app/src/main/java/org/stepic/droid/ui/fragments/CourseSearchFragment.java
package org.stepic.droid.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import org.stepic.droid.R;
import org.stepic.droid.base.App;
import org.stepic.droid.core.presenters.SearchCoursesPresenter;
import org.stepic.droid.storage.operations.Table;
import org.stepic.droid.ui.custom.AutoCompleteSearchView;
import org.stepic.droid.ui.util.SearchViewHelper;
import org.stepic.droid.ui.util.ToolbarHelperKt;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;

public class CourseSearchFragment extends CourseListFragmentBase {

    private final static String QUERY_KEY = "query_key";

    public static Fragment newInstance(String query) {
        Fragment fragment = new CourseSearchFragment();
        Bundle bundle = new Bundle();
        bundle.putString(QUERY_KEY, query);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Nullable
    private String searchQuery;

    @Inject
    SearchCoursesPresenter searchCoursesPresenter;

    private CompositeDisposable searchSuggestionsDisposable;

    @Override
    protected void injectComponent() {
        App.Companion
                .componentManager()
                .courseGeneralComponent()
                .courseListComponentBuilder()
                .build()
                .inject(this);
    }

    @Override
    protected void onReleaseComponent() {
        App.Companion
                .componentManager()
                .releaseCourseGeneralComponent();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setHasOptionsMenu(true);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        searchQuery = getArguments().getString(QUERY_KEY);
        super.onViewCreated(view, savedInstanceState);
        ToolbarHelperKt.initCenteredToolbar(this, R.string.search_title, true);
        emptySearch.setClickable(false);
        emptySearch.setFocusable(false);
        searchCoursesPresenter.attachView(this);
        searchCoursesPresenter.restoreState();
        swipeRefreshLayout.post(new Runnable() {
            @Override
            public void run() {
                searchCoursesPresenter.downloadData(searchQuery);
            }
        });

    }

    @Override
    public void onDestroyView() {
        searchCoursesPresenter.detachView(this);
        super.onDestroyView();
    }

    @Override
    protected Table getCourseType() {
        return null;
    }

    @Override
    public void showEmptyScreen(boolean isShowed) {
        if (isShowed) {
            emptySearch.setVisibility(View.VISIBLE);
            swipeRefreshLayout.setVisibility(View.GONE);
        } else {
            emptySearch.setVisibility(View.GONE);
            swipeRefreshLayout.setVisibility(View.VISIBLE);

        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.search_menu, menu);
        MenuItem searchMenuItem = menu.findItem(R.id.action_search);
        AutoCompleteSearchView searchView = (AutoCompleteSearchView) searchMenuItem.getActionView();

        searchView.setCloseIconDrawableRes(getCloseIconDrawableRes());
        searchView.setSearchable(getActivity());
        searchView.initSuggestions(rootView);

        searchSuggestionsDisposable = SearchViewHelper.INSTANCE.setupSearchViewSuggestionsSources(searchView, api, databaseFacade, null);

        searchMenuItem.expandActionView();
        if (searchQuery != null) {
            searchView.setQuery(searchQuery, false);
        }
        searchView.clearFocus();

        searchMenuItem.setOnActionExpandListener(new MenuItem.OnActionExpandListener() {
            @Override
            public boolean onMenuItemActionExpand(MenuItem item) {
                return true;
            }

            @Override
            public boolean onMenuItemActionCollapse(MenuItem item) {
                getActivity().finish();
                return true;
            }
        });
    }

    @Override
    public void onDestroyOptionsMenu() {
        if (searchSuggestionsDisposable != null) {
            searchSuggestionsDisposable.dispose();
        }
        super.onDestroyOptionsMenu();
    }

    @Override
    protected void onNeedDownloadNextPage() {
        searchCoursesPresenter.downloadData(searchQuery);
    }

    @Override
    public void onRefresh() {
        searchCoursesPresenter.refreshData(searchQuery);
    }
}

File: app/src/main/java/org/stepic/droid/ui/fragments/CourseSearchFragment.kt
package org.stepic.droid.ui.fragments

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.widget.SearchView
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.view_catalog_search_toolbar.*
import kotlinx.android.synthetic.main.view_centered_toolbar.*
import org.stepic.droid.R
import org.stepic.droid.base.App
import org.stepic.droid.core.presenters.SearchCoursesPresenter
import org.stepic.droid.model.CourseListType
import org.stepic.droid.ui.custom.AutoCompleteSearchView
import org.stepic.droid.ui.util.initCenteredToolbar
import javax.inject.Inject

class CourseSearchFragment: CourseListFragmentBase() {
    companion object {
        private const val QUERY_KEY = "query_key"

        fun newInstance(query: String): Fragment {
            val fragment = CourseSearchFragment()
            val bundle = Bundle()
            bundle.putString(QUERY_KEY, query)
            fragment.arguments = bundle
            return fragment
        }

        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }
    }

    private var searchQuery: String? = null

    @Inject
    lateinit var searchCoursesPresenter: SearchCoursesPresenter

    lateinit var searchIcon: ImageView

    override fun injectComponent() {
        App
            .componentManager()
            .courseGeneralComponent()
            .courseListComponentBuilder()
            .build()
            .inject(this)
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        searchQuery = arguments?.getString(QUERY_KEY)
        super.onViewCreated(view, savedInstanceState)
        initCenteredToolbar(R.string.search_title, true)
        emptySearch.isClickable = false
        emptySearch.isFocusable = false
        searchCoursesPresenter.attachView(this)
        searchCoursesPresenter.restoreState()
        searchIcon = searchViewToolbar.findViewById(androidx.appcompat.R.id.search_mag_icon) as ImageView
        swipeRefreshLayout.post { searchCoursesPresenter.downloadData(searchQuery) }
        setupSearchBar()
    }

    override fun onDestroyView() {
        searchCoursesPresenter.detachView(this)
        super.onDestroyView()
    }

    override fun getCourseType(): CourseListType? = null

    public override fun showEmptyScreen(isShown: Boolean) {
        if (isShown) {
            emptySearch.visibility = View.VISIBLE
            swipeRefreshLayout.visibility = View.GONE
        } else {
            emptySearch.visibility = View.GONE
            swipeRefreshLayout.visibility = View.VISIBLE
        }
    }

    override fun onNeedDownloadNextPage() {
        searchCoursesPresenter.downloadData(searchQuery)
    }

    override fun onRefresh() {
        searchCoursesPresenter.refreshData(searchQuery)
    }

    private fun setupSearchBar() {
        centeredToolbar.isVisible = false
        backIcon.isVisible = true
        if (android.os.Build.VERSION.SDK_INT < 21) {
            toolbarShadow.isVisible = true
        }
        searchViewToolbar.isVisible = true
        searchViewContainer.setBackgroundColor(ContextCompat.getColor(requireContext(), R.color.white))
        searchViewToolbar.setBackgroundColor(ContextCompat.getColor(requireContext(), R.color.white))
        searchIcon.setImageResource(0)
        (searchViewToolbar.layoutParams as ViewGroup.MarginLayoutParams).setMargins(0, 0, 0, 0)
        setupSearchView(searchViewToolbar)
        searchViewToolbar.setIconifiedByDefault(false)
        backIcon.setOnClickListener {
            val hasFocus = searchViewToolbar.hasFocus()
            if (hasFocus) {
                searchViewToolbar.clearFocus()
            } else {
                activity?.finish()
            }
        }
    }

    private fun setupSearchView(searchView: AutoCompleteSearchView) {
        searchView.setCloseIconDrawableRes(getCloseIconDrawableRes())
        searchView.setSearchable(requireActivity())
        searchView.initSuggestions(rootView)

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                searchView.onSubmitted(query)
                return false
            }

            override fun onQueryTextChange(query: String): Boolean {
                searchView.setConstraint(query)
                return false
            }
        })

        searchView.onActionViewExpanded()
        searchQuery?.let { searchView.setQuery(it, false) }
        searchView.clearFocus()
    }
}